### PR TITLE
[Kernel] Support reading multipart checkpoints

### PR DESCRIPTION
**Pull request by allisonport-db**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [x] Kernel
- [ ] Other (fill in here)

## Description

Support reading multipart checkpoints.

## How was this patch tested?

Adds a few unit tests and an end-to-end test.

I'd love to add more tests, but this is where it is starting to seem cumbersome that we can only read the last version? This means we need a different golden table for each test case, instead of being able to test each version of a table.

### CODE REVIEW

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java
Comment (allisonport-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java:1]: Copied from https://github.com/delta-io/delta/blob/master/spark/src/test/scala/org/apache/spark/sql/delta/FileNamesSuite.scala

File: kernel/kernel-api/src/test/scala/io/delta/kernel/CheckpointInstanceSuite.scala
Comment (allisonport-db) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/CheckpointInstanceSuite.scala:1]: Copied from https://github.com/delta-io/delta/blob/master/spark/src/test/scala/org/apache/spark/sql/delta/CheckpointInstanceSuite.scala

File: kernel/kernel-api/src/test/scala/io/delta/kernel/FileNamesSuite.scala
Comment (allisonport-db) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/FileNamesSuite.scala:1]: Copied from https://github.com/delta-io/delta/blob/master/spark/src/test/scala/org/apache/spark/sql/delta/FileNamesSuite.scala

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java [hunk 1]
```diff
@@ -30,13 +31,25 @@ public class CheckpointInstance
...
      */
-    public static final CheckpointInstance MAX_VALUE = new CheckpointInstance(-1);
+    public static final CheckpointInstance MAX_VALUE = new CheckpointInstance(Long.MAX_VALUE);
 
     public final long version;
     public final Optional<Integer> numParts;
 
     public CheckpointInstance(Path path) {
-        this(FileNames.getFileVersion(path));
+        String[] pathParts = path.getName().split("\\.");
```
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java:40]: Is there way to work without using the `Path`. Currently, the `Path` class is a copy of the Hadoop `Path` and it is just a lot of code which we don't need.

File: kernel/kernel-api/src/test/scala/io/delta/kernel/CheckpointInstanceSuite.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
```
Comment (vkorukanti) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/CheckpointInstanceSuite.scala:2]: nit: update year.

**Review by vkorukanti — COMMENTED** [review]
Changes to support the multi-part checkpoints LGTM.

Are there any unit tests around SnapshotManager in Delta-Spark or Standalone, that we can copy to have better test coverage?

If there are none, can we mock the `FileSystemClient` in`TableClient` argument passed to the `SnapshotManager` and test?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java [hunk 3]
```diff
@@ -30,13 +31,25 @@ public class CheckpointInstance
...
+
+        if (pathParts.length == 3 && pathParts[1].equals("checkpoint") &&
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java:42]: For each of these if branches, i'd comment an example to give a visual explanation of what is happening.
e.g. `000123.checkpoint.parquet`.

Also, for `00000000000000000010.checkpoint.0000000002.0000000003.parquet` do we not capture the `002`? We only save the version and the numParts?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java [hunk 4]
```diff
@@ -71,11 +78,18 @@ public List<Path> getCorrespondingFiles(Path path) {
                 Collections.singletonList(FileNames.checkpointFileSingular(path, version)));
     }
 
+    /**
+     * Comparison rules:
+     * 1. A [[CheckpointInstance]] with higher version is greater than the one with lower version.
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java:83]: `[[CheckpointInstance]]` is scala-style, not java. Just say `CheckpointInstance` or `{@link CheckpointInstance}`

**Review by scottsand-db — APPROVED** [review]
LGTM after addressing minor comments

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java (hunk 1 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java:40]: I think no reason to use Path here. We convert from String to Path in `SnapshotManager`. Will update

**Comment by allisonport-db** [comment]
> Are there any unit tests around SnapshotManager in Delta-Spark or Standalone, that we can copy to have better test coverage?
> 
> If there are none, can we mock the `FileSystemClient` in`TableClient` argument passed to the `SnapshotManager` and test?

All we have is [SnapshotManagementSuite](https://github.com/delta-io/delta/blob/master/spark/src/test/scala/org/apache/spark/sql/delta/SnapshotManagementSuite.scala) which mostly only tests recovery for corrupt checkpoints (which is not yet implemented.)

I'll see what can be copied over or added.

**Comment by allisonport-db** [comment]
@vkorukanti added tests related specifically to the CheckpointInstance and multi-part checkpoints. I think we could add more mocked tests specifically for some of SnapshotManager (i.e. getLogSegmentForVersion) but I'd prefer to do that separately as the code is complicated (maybe can be cleaned up...) and we don't have existing tests for it in standalone/spark

**Review by vkorukanti — APPROVED** [review]

**Pull request merged**